lgc/patch/PatchReadFirstLane.cpp

```diff
@@ -0,0 +1,245 @@
+        initialReadFirstLanes.insert(inst);
+        readFirstLaneBlocks.insert(inst->getParent());
+      }
+    }
+  }
+  bool changed = false;
+
+  // Try to lift readfirstlanes per block
+  for (auto curBb : readFirstLaneBlocks) {
+    // Collect the instructions that can be marked as uniform
+    DenseSet<Instruction *> markedUniforms(initialReadFirstLanes);
+    // Collect the instructions where the readfirstlanes can be lifted to
+    DenseSet<Instruction *> liftInsts;
+
+    // Reversely iterate over the block to find qualified instructions to lift readfirstlanes
+    for (BasicBlock::reverse_iterator instIt = curBb->rbegin(), endIt = curBb->rend(); instIt != endIt; ++instIt) {
+      Instruction *candidateInst = &*instIt;
+      // The candidate instrction for a lift should be marked as uniform
+      if (markedUniforms.find(candidateInst) != markedUniforms.end()) {
+        unsigned numOps = candidateInst->getNumOperands();
+        if (dyn_cast<CallInst>(candidateInst))
+          --numOps;
+
+        // Stop propagation when it is not worth lifting a readfirstlane
+        if (numOps > 2 || (numOps == 2 && ((m_divergenceAnalysis->isDivergent(candidateInst->getOperand(0)) &&
+                                            m_divergenceAnalysis->isDivergent(candidateInst->getOperand(1)))))) {
+          liftInsts.insert(candidateInst);
+          continue;
+        }
+
+        // Check each operand if it can be marked as uniform or can be qualified as a lift instruction
+        for (unsigned idx = 0; idx < numOps; ++idx) {
+          Value *operand = candidateInst->getOperand(idx);
+          if (auto operandInst = dyn_cast<Instruction>(operand)) {
+            // A readfirstlane lift should occur in one block
+            // Stop propagtion if one of the operands isn't in the same block
+            if (operandInst->getParent() != curBb) {
+              liftInsts.insert(candidateInst);
+              break;
+            }
+
+            // Skip the marked instuction
+            if (markedUniforms.find(operandInst) != markedUniforms.end())
+              continue;
+
+            // Check the operand that cannot be deduced as uniform by divergence analysis
+            if (!m_divergenceAnalysis->isUniform(operandInst)) {
+              // The qualified intruction must has all users marked as uniform
+              bool allUsersUniform = true;
+              for (auto user : operandInst->users()) {
+                auto userInst = dyn_cast<Instruction>(user);
+                if (markedUniforms.find(userInst) == markedUniforms.end()) {
+                  allUsersUniform = false;
+                  break;
+                }
+              }
+              if (allUsersUniform) {
+                auto loadInst = dyn_cast<LoadInst>(operandInst);
+                if (dyn_cast<CallInst>(operandInst) || (loadInst && loadInst->getType()->isIntegerTy(32))) {
+                  // Stop propagation
+                  liftInsts.insert(operandInst);
```

Comment:
I don't understand why you stop propagation here.

Reply:
I think the index may from a function call or a loadInst. So I set the them as the stop condition.